bpf/cilium-probe-kernel-hz.c

```diff
@@ -33,13 +33,26 @@ static int pin_to_cpu(int cpu)
 	return sched_setaffinity(0, sizeof(set), &set);
 }
 
-static int fix_priority(void)
+static int pin_scheduler(const int policy)
 {
 	struct sched_param sp = {
-		.sched_priority = sched_get_priority_max(SCHED_FIFO),
+		.sched_priority = sched_get_priority_max(policy),
 	};
 
-	return sched_setscheduler(0, SCHED_FIFO, &sp);
+	return sched_setscheduler(0, policy, &sp);
+}
+
+static void fix_priority(void)
+{
+	int ret_fifo = pin_scheduler(SCHED_FIFO);
+
+	if (ret_fifo) {
+		int ret_nice = setpriority(PRIO_PROCESS, 0, -20);
+
+		if (ret_fifo || ret_nice)
+			printf("/* Note, could not set %s for probing KERNEL_HZ. */\n",
+			       ret_fifo && ret_nice ? "fifo and nice" : "fifo");
```

Comment:
I tested this again with artificial load (processes with high priority on same CPU). When we can't set both fifo and nice, the program fails a lot more often, most of the time with a `probed jiffies mismatch` but also sometimes just with an incorrect HZ value (e.g., 300 or 1000).

I don't know what impact an incorrect HZ value would have on the datapath, but maybe we could also add a threshold on the warp value (when the HZ value is incorrect, warp is always 2-3-digits long)? Then again, I had to have several processes with highest priority running on the same core to see that happening, so probably quite unlikely.

Reply:
Ok, to be on safe side, I've reworked everything. Will post in a bit.